### DISCUSSION TITLE
Bookmark of PDF file

### DISCUSSION THREAD

**Discussion by Carnegie-eng** [pymupdf/PyMuPDF#2852] [Q&A] [answered]
Hi, I am trying to get and then edit bookmarks of a PDF file.
I confirmed the "outline" of the PyMuPDF document (https://pymupdf.readthedocs.io/en/latest/outline.html#outline), but I could not do the all the bookmark title of the PDF document. The PDF have several levels of bookmark. 
I appreciate if you could teach me a example code. Thank you so much for your support !

**Comment by JorjMcKie** [comment, marked as answer]
Fortunately, dealing with Tables of Content (TOC) is very easy in PyMuPDF. You almost never are required to actually look at the [outline](https://pymupdf.readthedocs.io/en/latest/outline.html#outline) object.

Simply prepare a list of sub-lists. Each sub-list must have this layout: `[level, title, page]`. Please also consult the documentation [here](https://pymupdf.readthedocs.io/en/latest/document.html#Document.set_toc). A simple update session is shown [here](https://pymupdf.readthedocs.io/en/latest/document.html#set-toc-demonstration).

**Comment by Carnegie-eng** [comment]
Thank you so much. It works! PyMuPDF package is wonderful !
